src/connection.rs

```diff
@@ -1,241 +1,54 @@
-use std::net::TcpStream;
-use std::ops::{Deref, DerefMut};
-#[cfg(unix)]
-use std::os::unix::net::UnixStream;
-use url::Url;
-
-use crate::error::MemcacheError;
-
-use crate::protocol::{AsciiProtocol, ProtocolTrait};
-use crate::stream::Stream;
-#[cfg(feature = "tls")]
-use openssl::ssl::{SslConnector, SslFiletype, SslMethod, SslVerifyMode};
-use r2d2::ManageConnection;
-
-/// A connection to the memcached server
-#[allow(missing_debug_implementations)]
-pub struct Connection(AsciiProtocol<Stream>);
-
-impl DerefMut for Connection {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
-    }
-}
-
-impl Deref for Connection {
-    type Target = AsciiProtocol<Stream>;
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
-
-/// Memcache connection manager implementing rd2d Pool ManageConnection
+use pin_project::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::io::{AsyncBufRead, AsyncRead, AsyncWrite, BufReader, BufWriter, ReadBuf};
+use tokio::net::{TcpStream, ToSocketAddrs};
+
+/// Connection wrapper
 #[derive(Debug)]
-pub struct ConnectionManager {
-    url: Url,
-}
+#[pin_project]
+pub struct Connection(#[pin] BufReader<BufWriter<TcpStream>>);
 
-impl ConnectionManager {
-    /// Initialize connection manager with given Url
-    pub fn new(target: impl AsRef<str>) -> Result<Self, MemcacheError> {
-        let url = Url::parse(target.as_ref())?;
-        Ok(Self { url })
+impl AsyncRead for Connection {
+    fn poll_read(
+        self: Pin<&mut Self>,
```

Comment:
Haven't seen this syntax, why would you need to specify type for self? 🤔

Reply:
this is because it is projected, maybe this would help https://doc.rust-lang.org/std/pin/index.html

Reply:
It does, thank you!